docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,108 @@
+---
+title: "Cross-Tenant Query Federation"
+linkTitle: "Cross-Tenant Query Federation"
+weight: 1
+slug: "cross-tenant-query-federation"
+---
+
+- Author: [Christian Simon](https://github.com/simonswine)
+- Date: October 2020
+- Status: Proposed
+
+## Overview
+
+This document aims to describe how to implement the ability to allow queries to cover data from more than a single Cortex tenant.
+
+## Reasoning
+
+Adopting a tenancy model within an organization with each tenant representing a department comes with the disadvantage that it will prevent queries from spanning multiple departments. This proposal tries to overcome those limitations.
+
+## Alternatives considered
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to be able to always identify the tenant correctly, queries using multiple tenants should inject a tenant label named `__tenant__` and its value containing the tenant ID into the results. A potentially existing label with the same name should be stored in a label with the prefix `original_`.
+
+Label selectors containing the tenant label should behave like any other labels. This can be achieved by selecting the tenants used in a multi tenant query.
+
+### Exposure of feature to the user
+
+#### Challenge
+
+The tenant ID is currently read from the `X-Scope-OrgID` HTTP header. The tenant ID has those [documented limitations][cortex_tenant_id] of values being allowed.
+
+[cortex_tenant_id]: https://cortexmetrics.io/docs/guides/limitations/#tenant-id-naming
+
+#### Proposal
+
+For the query path a user should be able to specify a `X-Scope-OrgID` header with multiple tenant IDs.  Multiple tenant IDs should then be propagating throughout out the system until it reaches the querier. The `Queryable` instance returned by the querier component, is wrapped by a `mergeQueryable`, which will aggregate the results from a `Queryable` per tenant and hence treated by the downstream components as a single tenant query.
+
+To allow such queries to be processed we suggest that an experimental configuration flag `-experimental.querier.tenant-federation.enabled` will be added, which is switched off by default. Once enabled the value of the `X-Scope-OrgID` header should be interpreted as `|` separated list of tenant ids. Components which are not expecting multiple tenant ids (e.g. the ingress path) must signal an error if multiple are used.
```

Comment:
Please don't use the ` --experimental` prefix. We're not going to do it anymore, because it's a pain to remove it once the feature is stable. Experimental features are just listed to `docs/configuration/v1-guarantees.md`, so the CLI flag name can just be `-querier.tenant-federation.enabled`.

Reply:
Makes sense. Added this change.